### PR TITLE
feat(tabs): remove internal support for history and url manipulation - FE-3484

### PR DESCRIPTION
**Pull request by bulaj**
closes: #3520

### Proposed behaviour
Removes baked in support for manipulation of window location as implementation is extremely limited and implementing the a feature to override it will be difficult to maintain. Consumers wishing to have this functionality should utilise the `onTabChange` and `selectedTabId` props and implement their own `history` implementation.

BREAKING CHANGE: Internal support for updating window location has been removed from the `Tabs`
component, the `setLocation` prop has also been removed as it serves no purpose beyond serving as a
guard for this functionality. Consumers who wish to have this functionality should utilise the
`onTabChange` and `selectedTabId` props with whatever `history` implementation they want to.

Removes dark background from focused `TabTitle` when `isTabSelected` is false 

### Current behaviour
There is no ability to pass external history to Tabs component.
 The current implementation only replaces and only uses hash history when `setLocation` prop is true.

### Checklist

- [x] Commits follow our style guide
- [ ] <del>Screenshots are included in the PR if useful
- [ ] <del>All themes are supported if required
- [x] Unit tests added or updated if required
- [ ] <del>Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [ ] <del>Carbon implementation and Design System documentation are congruent

### Testing instructions

Needs to be tested in external project containing react-router. See #3520.

https://codesandbox.io/s/demo-external-history-and-tabs-lwk3i

### CODE REVIEW

File: docs/usage-with-routing.stories.mdx
Comment (sjohnson-sage) [inline docs/usage-with-routing.stories.mdx:26]: The import path should have `/tabs` with a lower case `t` and gaps in this bit `{ Tabs, Tab }`

Comment (sjohnson-sage) [reply docs/usage-with-routing.stories.mdx:26]: I recommend setting up prettier to run on save in vscode (or whatever editor you use)

File: docs/usage-with-routing.stories.mdx
Comment (sjohnson-sage) [inline docs/usage-with-routing.stories.mdx:117]: Same as above

File: src/components/tabs/tabs.d.ts
Comment (sjohnson-sage) [inline src/components/tabs/tabs.d.ts:10]: This should be marked as optional

**Review by sjohnson-sage — COMMENTED** [review]
Just some minor changes to make 👍🏻

**Review by edleeks87 — COMMENTED** [review]
Nothing more to add that Sam hasn't already raised, happy to approve once they're addressed 👍

File: src/components/tabs/tabs.component.js
Comment (jamime) [inline src/components/tabs/tabs.component.js:48]: We shouldn't be defining things on the window.

**Review by jamime — CHANGES_REQUESTED** [review]
We should be using the `history` API instead of overriding `window.history`. I think it's acceptable to to use `history@5` as a dependency of our project so we can remove the `window.history.pushState`. However, it should still be a prop so users can pass their own `history` instance we should use the `history@5` API.

> We need to ensure that we can use react-router@5 with history@4 and react-router@6 with history@5.

This doesn't have to be done in the component as long as the user can achieve the desired result e.g. by writing a wrapper around `history@4` so it functions the same as `history@5`.

[`history@4` documentation](https://github.com/ReactTraining/history/blob/3f69f9e07b0a739419704cffc3b3563133281548/docs/Navigation.md)

[`history@5` documentation](https://github.com/ReactTraining/history/blob/master/docs/navigation.md)

**Review by edleeks87 — APPROVED** [review]

**Review by nicktitchmarsh — DISMISSED** [review]

**Review by edleeks87 — DISMISSED** [review]

File: src/components/tabs/tabs.component.js
Comment (jamime) [inline src/components/tabs/tabs.component.js:128]: This doesn't look correct, the method shouldn't change based on `isExternalHistory`.

I think we need an additional prop that defaults to `replace`, the existing behaviour.
```
navigatorMethod: PropTypes.oneOf(['push', 'replace']);
```
then we would call
```js
historyAPI[navigatorMethod](url);
```

File: docs/usage-with-routing.stories.mdx
Comment (jamime) [inline docs/usage-with-routing.stories.mdx:58]: this will cause `history.push` to be called twice based on the code in `tabs.component.js`, I think you also need to set `setLocation=false`.

File: docs/usage-with-routing.stories.mdx
Comment (jamime) [inline docs/usage-with-routing.stories.mdx:145]:
```suggestion
    <Tabs history={history} onTabChange={handleTabChange} setLocation={false}>
```

**Review by jamime — CHANGES_REQUESTED** [review]
> There is no code in Carbon to implement a specific routing library, we aim to let the user decide which library to use.

This is no longer true in the `usage-with-routing.stories.mdx`.

**Review by jamime — CHANGES_REQUESTED** [review]
I've left some comments, I think we need a few possible combinations

- default history push
- default history replace
- custom history push
- custom history replace
- custom history onTabChange

I've created a sandbox to show these proposed changes

https://codesandbox.io/s/3865-tabs-tip5c

**Review by edleeks87 — APPROVED** [review]

File: src/components/tabs/tabs.component.js
Comment (jamime) [inline src/components/tabs/tabs.component.js:36]:
```suggestion
```

File: package.json
Comment (jamime) [inline package.json:114]:
```suggestion
```

**Review by jamime — CHANGES_REQUESTED** [review]
Just some small leftovers from previous implementation.

File: src/components/tabs/tabs.component.js
Comment (edleeks87) [reply src/components/tabs/tabs.component.js:36]: 🤦

File: src/components/tabs/tabs.component.js
Comment (sjohnson-sage) [inline src/components/tabs/tabs.component.js:50]: This check using `firstRender` could be replaced by an empty dependency array so it only runs on the first render.

Don't have to change it here because it's not part of your changes, but if you want to tidy that up it's up to you

File: docs/usage-with-routing.stories.mdx
Comment (sjohnson-sage) [inline docs/usage-with-routing.stories.mdx:57]: We need a `selectedTabId` on this as well don't we?

File: docs/usage-with-routing.stories.mdx
Comment (sjohnson-sage) [inline docs/usage-with-routing.stories.mdx:139]: Same as above

File: src/components/tabs/tabs.stories.mdx
Comment (edleeks87) [reply src/components/tabs/tabs.stories.mdx:1605]: Yeah agreed, it was the linter complaining about it being a dependency in the useEffect that made me useMemo, but I'll just tidy the implementation up and remove it

File: src/components/tabs/tabs.component.js
Comment (edleeks87) [reply src/components/tabs/tabs.component.js:368]: Yeah again not sure how i missed that 🤦

File: src/components/tabs/tabs.component.js
Comment (edleeks87) [reply src/components/tabs/tabs.component.js:50]: yeah good point I'll update it now the dependencies have been reduced

File: docs/usage-with-routing.stories.mdx
Comment (edleeks87) [reply docs/usage-with-routing.stories.mdx:57]: Yep, I'm on fire 🔥  with this Pr

File: src/components/tabs/tabs.component.js
Comment (sjohnson-sage) [reply src/components/tabs/tabs.component.js:50]: The dependencies are irrelevant because we only want this stuff to run on first render, they're only there to trigger re-runs of the function when the dependencies change

You might have to disable an eslint rule to have the empty array but it's the right thing to do here!

Comment (edleeks87) [reply src/components/tabs/tabs.component.js]: Yeah sorry I meant they were reduced down from what they were previously so it's worth updating the layoutEffect further rather than leaving it as is for now etc

File: docs/usage-with-routing.stories.mdx
Comment (sjohnson-sage) [inline docs/usage-with-routing.stories.mdx:107]: The example below doesn't use `createBrowserHistory` anywhere, I think you can get rid of this line

File: src/components/tabs/tabs.component.js
Comment (sjohnson-sage) [inline src/components/tabs/tabs.component.js:51]: What's the point of the `.filter((child) => child)` bit here? This could just be `React.Children.toArray(children)[0].props.tabId;` I think

File: src/components/tabs/tabs.stories.mdx
Comment (sjohnson-sage) [reply src/components/tabs/tabs.stories.mdx:1607]: This is still here btw

File: src/components/tabs/tabs.stories.mdx
Comment (sjohnson-sage) [inline src/components/tabs/tabs.stories.mdx:1593]: `manipulate`

File: src/components/tabs/tabs.component.js
Comment (edleeks87) [reply src/components/tabs/tabs.component.js:51]: Yeah I added it to account for conditional rendering of the `Tab` `children` but I think you're right and it's overkill so I'll update it

**Review by sjohnson-sage — DISMISSED** [review]

**Review by jamime — DISMISSED** [review]

**Review by sjohnson-sage — DISMISSED** [review]

**Review by jamime — COMMENTED** [review]
LGTM please squash commits before merging.

**Review by sjohnson-sage — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Pull request merged**
